src/TraceEvent/EventPipe/EventPipeEventSource.cs

```diff
@@ -58,80 +90,89 @@ protected override void Dispose(bool disposing)
             {
-                TraceEventNativeMethods.EVENT_RECORD* eventRecord = ReadEvent(deserializerReader);
-                if (eventRecord != null)
+                // loop through the stream until we hit a null object.  Deserialization of 
+                // EventPipeEventBlocks will cause dispatch to happen.  
+                // ReadObject uses registered factories and recognizes types by names, then derserializes them with FromStream
+                while (_deserializer.ReadObject() != null)
+                { }
+            }
+#if SUPPORT_V1_V2
+            else
+            {
+                PinnedStreamReader deserializerReader = (PinnedStreamReader)_deserializer.Reader;
+                while (deserializerReader.Current < _endOfEventStream)
                 {
-                    // in the code below we set sessionEndTimeQPC to be the timestamp of the last event.  
-                    // Thus the new timestamp should be later, and not more than 1 day later.  
-                    Debug.Assert(sessionEndTimeQPC <= eventRecord->EventHeader.TimeStamp);
-                    Debug.Assert(sessionEndTimeQPC == 0 || eventRecord->EventHeader.TimeStamp - sessionEndTimeQPC < _QPCFreq * 24 * 3600);
-
-                    TraceEvent event_ = Lookup(eventRecord);
-                    Dispatch(event_);
-                    sessionEndTimeQPC = eventRecord->EventHeader.TimeStamp;
+                    TraceEventNativeMethods.EVENT_RECORD* eventRecord = ReadEvent(deserializerReader);
+                    if (eventRecord != null)
+                    {
+                        // in the code below we set sessionEndTimeQPC to be the timestamp of the last event.  
+                        // Thus the new timestamp should be later, and not more than 1 day later.  
+                        Debug.Assert(sessionEndTimeQPC <= eventRecord->EventHeader.TimeStamp);
+                        Debug.Assert(sessionEndTimeQPC == 0 || eventRecord->EventHeader.TimeStamp - sessionEndTimeQPC < _QPCFreq * 24 * 3600);
+
+                        var traceEvent = Lookup(eventRecord);
+                        Dispatch(traceEvent);
+                        sessionEndTimeQPC = eventRecord->EventHeader.TimeStamp;
+                    }
                 }
             }
-
+#endif
             return true;
         }
 
-        internal override string ProcessName(int processID, long timeQPC)
-        {
-            return _processName;
-        }
+        internal override string ProcessName(int processID, long timeQPC) => _processName;
 
-        private TraceEventNativeMethods.EVENT_RECORD* ReadEvent(PinnedStreamReader reader)
+        internal TraceEventNativeMethods.EVENT_RECORD* ReadEvent(PinnedStreamReader reader)
         {
-            // Guess that the event is < 1000 bytes or whatever is left in the stream.  
-            int eventSizeGuess = Math.Min(1000, _endOfEventStream.Sub(reader.Current));
+            nint eventSizeGuess =
```

Comment:
As a rule I don't like to use ?: except in cases where it clearly fits on a line.  For everything else use if statements. 

However I think I was trying to be too smart and just made the code complex.   Time  to simplify.   For the first call to Get Pointer simply use EventPipeHeaderSize.   Then make the call to GetPointer down below unconditional.   This is simpler and probably as efficient as what is there now.